extensions/fetch/20_headers.js

```diff
@@ -178,32 +178,42 @@
     get [_iterableHeaders]() {
       const list = this[_headerList];
 
-      const headers = [];
-      const headerNamesSet = new Set();
+      // The order of steps are not similar to the ones suggested by the
+      // spec but produce the same result.
+      const headers = {};
+      const cookies = [];
       for (const entry of list) {
-        headerNamesSet.add(byteLowerCase(entry[0]));
-      }
-      const names = [...headerNamesSet].sort();
-      for (const name of names) {
-        // The following if statement, and if block of the following statement
-        // are not spec compliant. `set-cookie` is the only header that can not
-        // be concatentated, so must be given to the user as multiple headers.
+        const name = byteLowerCase(entry[0]);
+        const value = entry[1];
+        if (value === null) throw new TypeError("Unreachable");
+        // The following if statement is not spec compliant.
+        // `set-cookie` is the only header that can not be concatentated,
+        // so must be given to the user as multiple headers.
         // The else block of the if statement is spec compliant again.
-        if (name == "set-cookie") {
-          const setCookie = list.filter((entry) =>
-            byteLowerCase(entry[0]) === "set-cookie"
-          );
-          if (setCookie.length === 0) throw new TypeError("Unreachable");
-          for (const entry of setCookie) {
-            headers.push([name, entry[1]]);
-          }
+        if (name === "set-cookie") {
+          cookies.push([name, value]);
         } else {
-          const value = getHeader(list, name);
-          if (value === null) throw new TypeError("Unreachable");
-          headers.push([name, value]);
+          // The below if/else block has the same behaviour as getHeaders()
+          // at the end of loop. But it avoids looping through the entire
+          // list to combine multiple values with same header name. It
+          // instead gradually combines them as they are found.
+          if (headers[name] && headers[name].length > 0) {
+            const oldValues = headers[name].split("\x2C\x20");
+            oldValues.push(value);
+            headers[name] = oldValues.join("\x2C\x20");
+          } else {
+            headers[name] = value;
+          }
         }
       }
-      return headers;
+
+      return [...Object.entries(headers), ...cookies].sort((a, b) => {
```

Comment:
I confess ignorance: does the spec require alphasort?

Reply:
Yes: https://fetch.spec.whatwg.org/#concept-header-list-sort-and-combine -> https://fetch.spec.whatwg.org/#convert-header-names-to-a-sorted-lowercase-set

Reply:
But the internal header list may not yet be sorted, because the original sort order is important for the order of headers on the wire.

Reply:
Thanks, duly noted. What a weird requirement though. Do you happen to know what the rationale is? I'm assuming there is one.

Reply:
No idea. Maybe to not leak the ordering of headers on the wire into user land? No idea though.